### PR TITLE
feat(DTFS2-7550): add cron job to schedule cancellations

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
When a deal is cancelled with an effective date in the future, a cron job needs to update it once this comes to pass

## Resolution :heavy_check_mark:
- Add `findScheduledDealCancellations` method to `TfmDealCancellationRepo`
- Add `cancelDealJob` cron job to find scheduled cancellations from the past & update the deal to cancelled

### CODE REVIEW

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { TfmDealRepo } from '../../repositories/tfm-deals-repo/tfm-deals.repo';
+
+/**
+ * Updates deals to be cancelled
+ * @param dealIds the deals to be cancelled
+ */
+const cancelDeals = async (dealIds: ObjectId[]) => {
+  await Promise.all(
+    dealIds.map((dealId) =>
+      TfmDealRepo.updateOneDeal(
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:15]: Replace this with `TfmDealCancelaltionRepo.submitDealCancellation`

File: dtfs-central-api/src/repositories/tfm-deals-repo/tfm-deals.repo.ts
Comment (MarRobSoftwire) [inline dtfs-central-api/src/repositories/tfm-deals-repo/tfm-deals.repo.ts:1]: Delete this file

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts (hunk 1 above)
Comment (MarRobSoftwire) [reply dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:15]: From #3696

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import { TfmDealCancellationRepo } from '../../repositories/tfm-deals-repo';
+
+const { DEAL_CANCELLATION_SCHEDULE } = process.env;
+
+/**
+ * Updates deals to be cancelled
+ * @param dealIds the deals to be cancelled
+ */
+const cancelDeals = async (deals: TfmDeal[]) => {
+  await Promise.all(deals.map((deal) => TfmDealCancellationRepo.submitDealCancellation(deal._id, deal.tfm.cancellation!, generateSystemAuditDetails())));
```
Comment (abhi-markan) [inline dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:13]: Just wondering if we need to add `non-null assertion` operator?

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+};
+
+/**
+ * Gets deal ids with a scheduled cancellation & effective date today or in the past
+ * @returns The deal ids to cancel
+ */
+const getDealsWithCancellationsScheduledForThePast = async (): Promise<TfmDeal[]> => {
+  const dealsScheduledForCancellation = await TfmDealCancellationRepo.findScheduledDealCancellations();
+
+  return dealsScheduledForCancellation.filter((deals) => deals.tfm.cancellation!.effectiveFrom < endOfDay(new Date()).valueOf());
```
Comment (abhi-markan) [inline dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:23]: Same as above, wondering if we should rather use optional chaining?

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+};
+
+/**
+ * Cancels deals scheduled to be cancelled in the past
+ */
+const cancelScheduledDeals = async (): Promise<void> => {
+  const deals = await getDealsWithCancellationsScheduledForThePast();
```
Comment (abhi-markan) [inline dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:30]:
```suggestion
  const deals = await getDealsWithPastCancellationsScheduled();
```

In the interest of keeping names short.

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts (hunk 2 above)
Comment (abhi-markan) [reply dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:13]: Should we check whether deal is already cancelled?

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts (hunk 3 above)
Comment (abhi-markan) [reply dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts:23]: Should we add a condition to make we are only fetching deals which are not cancelled?

File: dtfs-central-api/src/repositories/tfm-deals-repo/tfm-deal-cancellation.repo.ts [hunk 5]
```diff
@@ -48,6 +48,21 @@ export class TfmDealCancellationRepo {
...
+   * Find deals with scheduled cancellations
+   * @returns the deals
+   */
+  public static async findScheduledDealCancellations(): Promise<TfmDeal[]> {
+    const dealCollection = await this.getCollection();
+
+    return await dealCollection
+      .find({
+        'dealSnapshot.submissionType': { $in: [DEAL_SUBMISSION_TYPE.AIN, DEAL_SUBMISSION_TYPE.MIN] },
+        'tfm.cancellation.status': { $eq: TFM_DEAL_CANCELLATION_STATUS.SCHEDULED },
```
Comment (abhi-markan) [inline dtfs-central-api/src/repositories/tfm-deals-repo/tfm-deal-cancellation.repo.ts:61]: Should we add a condition to ensure `tfm.stage` is not cancelled?

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Few minor comments to address please.

**Review by abhi-markan — COMMENTED** [review]
Just a documentation comment thanks.

**Review by abhi-markan — APPROVED** [review]

File: dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts (hunk 2 above)
Comment (MarRobSoftwire) [reply dtfs-central-api/src/cron-scheduler-jobs/deal-cancellation/cancel-deal-job.ts]: resolving conversation since approved so can merge in :)

**Pull request merged**